vllm/distributed/kv_transfer/kv_connector/v1/nixl_connector/handshake.py

```diff
@@ -0,0 +1,279 @@
+                                 tp_rank: int):
+        encoder = msgspec.msgpack.Encoder()
+        encoded_data = encoder.encode(metadata)
+        size_in_bytes = len(encoded_data)
+        logger.debug("Size of encoded NixlAgentMetadata: %s bytes",
+                     size_in_bytes)
+
+        # Listen for new requests for metadata
+        host = envs.VLLM_NIXL_SIDE_CHANNEL_HOST
+        path = make_zmq_path("tcp", host, base_port + tp_rank)
+        logger.debug("Starting listening on path: %s", path)
+        with ZmqHandshakeStrategy._zmq_ctx(zmq.ROUTER, path) as sock:
+            ready_event.set()
+            while True:
+                identity, _, msg = sock.recv_multipart()
+                if msg != GET_META_MSG:
+                    logger.warning(
+                        "Connection listener got unexpected message %s", msg)
+                sock.send_multipart((identity, b"", encoded_data))
+
+    @staticmethod
+    @contextlib.contextmanager
+    def _zmq_ctx(socket_type: Any, addr: str) -> Iterator[zmq.Socket]:
+        if socket_type not in (zmq.ROUTER, zmq.REQ):
+            raise ValueError(f"Unexpected socket type: {socket_type}")
+
+        ctx: Optional[zmq.Context] = None
+        try:
+            ctx = zmq.Context()
+            yield make_zmq_socket(ctx=ctx,
+                                  path=addr,
+                                  socket_type=socket_type,
+                                  bind=socket_type == zmq.ROUTER)
+        finally:
+            if ctx is not None:
+                ctx.destroy(linger=0)
+
+
+class HttpHandshakeStrategy(HandshakeStrategy):
+    """
+    Handshake strategy that uses HTTP requests to fetch metadata from a
+    remote server. This is done through the front-end, and is
+    North-South, not P2P.
+    """
+
+    def __init__(self, nixl_wrapper, tp_rank: int, tp_size: int,
+                 side_channel_port: int, engine_id: str,
+                 add_remote_agent_func):
+        super().__init__(nixl_wrapper, tp_rank, tp_size, side_channel_port,
+                         engine_id)
+        self.add_remote_agent_func = add_remote_agent_func
+        self._tp_size: dict[str, int] = {engine_id: tp_size}
+
+    def initiate_handshake(self, host: str, port: int, remote_tp_size: int,
+                           expected_engine_id: str) -> dict[int, str]:
+        start_time = time.perf_counter()
+        logger.debug("Starting NIXL handshake with %s:%s", host, port)
+
+        url = build_uri("http", host, port, path="get_kv_connector_metadata")
+
+        try:
+            http_client = HTTPConnection()
+            response = http_client.get_response(
+                url, timeout=envs.VLLM_NIXL_HANDSHAKE_TIMEOUT)
+            response.raise_for_status()
+            res = response.json()
+        except Exception as e:
+            logger.error("Failed to fetch metadata from %s: %s", url, e)
+            raise
```

Comment:
I think we can raise from here

Reply:
I ended up de-indenting since we are calling raise_from_status(), it makes sense to throw from anywhere in this function since we want this to be externally retryable